contracts/samples/TokenPaymaster.sol

```diff
@@ -1,113 +1,213 @@
+            uint256 priceMarkup = tokenPaymasterConfig.priceMarkup;
+            uint256 paymasterAndDataLength = userOp.paymasterAndData.length - 20;
+            require(paymasterAndDataLength == 0 || paymasterAndDataLength == 32,
+                "TPM: invalid data length"
+            );
+            uint256 preChargeNative = requiredPreFund + (REFUND_POSTOP_COST * userOp.maxFeePerGas);
+        // note: as price is in ether-per-token and we want more tokens increasing it means dividing it by markup
+            uint256 cachedPriceWithMarkup = cachedPrice * PRICE_DENOMINATOR / priceMarkup;
+            if (paymasterAndDataLength == 32) {
+                uint256 clientSuppliedPrice = uint256(bytes32(userOp.paymasterAndData[20 : 52]));
+                if (clientSuppliedPrice < cachedPriceWithMarkup) {
+                    // note: smaller number means 'more ether per token'
+                    cachedPriceWithMarkup = clientSuppliedPrice;
+                }
+            }
+            uint256 tokenAmount = weiToToken(preChargeNative, cachedPriceWithMarkup);
+            SafeERC20.safeTransferFrom(token, userOp.sender, address(this), tokenAmount);
+            context = abi.encodePacked(tokenAmount, userOp.maxFeePerGas, userOp.sender);
+            validationResult = 0;
         }
+    }
 
-        return (abi.encode(userOp.sender), 0);
+    /// @notice Performs post-operation tasks, such as updating the token price and refunding excess tokens.
+    /// @dev This function is called after a user operation has been executed or reverted.
+    /// @param mode The post-operation mode (either successful or reverted).
+    /// @param context The context containing the token amount and user sender address.
+    /// @param actualGasCost The actual gas cost of the transaction.
+    function _postOp(PostOpMode mode, bytes calldata context, uint256 actualGasCost) internal override {
+        if (mode == PostOpMode.postOpReverted) {
+            return;
+        }
+        unchecked {
+            uint256 priceMarkup = tokenPaymasterConfig.priceMarkup;
+            uint256 preCharge = uint256(bytes32(context[0 : 32]));
+            uint256 maxFeePerGas = uint256(bytes32(context[32 : 64]));
+            address userOpSender = address(bytes20(context[64 : 84]));
+            if (mode == PostOpMode.postOpReverted) {
+                emit PostOpReverted(userOpSender, preCharge);
+                // Do nothing here to not revert the whole bundle and harm reputation
+                return;
+            }
+            uint256 _cachedPrice = updateCachedPrice(false);
+        // note: as price is in ether-per-token and we want more tokens increasing it means dividing it by markup
+            uint256 cachedPriceWithMarkup = _cachedPrice * PRICE_DENOMINATOR / priceMarkup;
+        // Refund tokens based on actual gas cost
+            uint256 actualChargeNative = actualGasCost + REFUND_POSTOP_COST * maxFeePerGas;
```

Comment:
instead of using maxFeePerGas, does it make sense to use tx.gasprice ?

Reply:
You are right, thanks! However, we cannot just use `tx.gasprice` as the bundler can manipulate this value. I think we should use `userOp.gasPrice()` here. Updated.

Reply:
userop.gasPrice() touches BASEFEE which forbidden opcode. this would be fine in postOp(), but you shouldn't calculate it in validatePaymasterUserOp and pass it in context.
Instead pass maxPriorityFee and maxFee in context and calculate like below in the postOp

```
uint256 opGasPrice;
        unchecked {
            if (maxFeePerGas == maxPriorityFeePerGas) {
                opGasPrice = maxFeePerGas;
            } else {
                opGasPrice = Math.min(maxFeePerGas, maxPriorityFeePerGas + block.basefee);
            }
        }
```

Reply:
Right 😀